http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java

```diff
@@ -162,6 +168,36 @@ private SslProvider resolveSslProvider(DefaultBuilder builder) {
         return SslContext.defaultClientProvider();
     }
 
+    private ProtocolNegotiation resolveProtocolNegotiation(ProtocolNegotiation userSetValue, AttributeMap serviceDefaultsMap,
+                                                           Protocol protocol) {
+        if (userSetValue == ProtocolNegotiation.ALPN) {
+            // TODO - remove once we implement support for ALPN with HTTP1
+            if (protocol == Protocol.HTTP1_1) {
+                throw new UnsupportedOperationException("ALPN with HTTP1 is not yet supported, use prior knowledge instead with "
+                                                        + "ProtocolNegotiation.ASSUME_PROTOCOL, or use ALPN with H2.");
+            }
+
+            // throw error if not supported and user set ALPN
+            validateAlpnSupported();
+            return ProtocolNegotiation.ALPN;
+        }
+        if (userSetValue == ProtocolNegotiation.ASSUME_PROTOCOL) {
+            return ProtocolNegotiation.ASSUME_PROTOCOL;
+        }
+
+        ProtocolNegotiation protocolNegotiation = serviceDefaultsMap.get(SdkHttpConfigurationOption.PROTOCOL_NEGOTIATION);
+        if (protocolNegotiation == ProtocolNegotiation.ALPN) {
+            if (!isAlpnSupported()) {
+                // fallback to prior knowledge if not supported and SDK defaults to ALPN
+                protocolNegotiation = ProtocolNegotiation.ASSUME_PROTOCOL;
+                log.warn(null, () -> "ALPN is not supported in the current Java version, falling back to prior knowledge for "
+                                     + "protocol negotiation");
+            }
```

Comment:
I think we should throw if the ALPN is set on the HTTP client builder and the JDK doesn't support it.

Reply:
This scenario is future new H2 service, async client, with old Java 8 version.So we want to fail by default and force user to upgrade Java version or explicitly set ASSUME_PROTOCOL on the Netty client?

Reply:
Oh I see, I thought this was an explicit setting on the builder, set by the customer. I think this is fine then. If the service doesn't support prior knowledge at least we log a warning.